File: app/src/main/java/com/example/acdat_pizzeria/modelo/Pizza.java
package com.example.acdat_pizzeria.modelo;

import android.content.ContentValues;

import com.example.acdat_pizzeria.R;
import com.example.acdat_pizzeria.daos.DAOPizzas;
import com.example.acdat_pizzeria.daos.data.PizzaEntry;
import com.example.acdat_pizzeria.daos.data.UsuarioEntry;
import com.example.acdat_pizzeria.modelo.enums.Ingrediente;
import com.example.acdat_pizzeria.modelo.enums.Queso;
import com.example.acdat_pizzeria.modelo.enums.Salsa;
import com.example.acdat_pizzeria.modelo.enums.Tamano;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Objects;

public class Pizza implements Serializable {

    private static Integer contId = 0;
    private Integer idPizza;
    private Tamano tamano;
    private Salsa salsa;
    private Queso queso;
    private ArrayList<Ingrediente> ingredientes;
    private Usuario usuario;
    private Boolean favorita;
    private String nombre;
    private int ref_img;

    public Pizza(Integer idPizza, Tamano tamano, Salsa salsa, Queso queso, ArrayList<Ingrediente> ingredientes, Usuario usuario, String nombre) {
        this.idPizza = idPizza;
        this.tamano = tamano;
        this.salsa = salsa;
        this.queso = queso;
        this.ingredientes = ingredientes;
        this.usuario = usuario;
        this.favorita = false;
        this.nombre = nombre;
        this.ref_img = R.drawable.pizza;
    }

    public Pizza(Tamano tamano, Salsa salsa, Queso queso, ArrayList<Ingrediente> ingredientes, Usuario usuario) {
        this.idPizza = incrmentId();
        this.tamano = tamano;
        this.salsa = salsa;
        this.queso = queso;
        this.ingredientes = ingredientes;
        this.usuario = usuario;
        this.favorita = false;
        this.nombre = "Predeterminada";
        this.ref_img = R.drawable.pizza;
    }

    public Pizza(Salsa salsa, Queso queso, ArrayList<Ingrediente> ingredientes, String nombre, int ref_img) {
        this.idPizza = incrmentId();
        this.salsa = salsa;
        this.queso = queso;
        this.ingredientes = ingredientes;
        this.favorita = false;
        this.nombre = nombre;
        this.ref_img = ref_img;
    }

    /*
    public Pizza(Tamano tamano, Salsa salsa, Queso queso, ArrayList<Ingrediente> ingredientes, Usuario usuario, String nombre) {
        this.idPizza = incrmentId();
        this.tamano = tamano;
        this.salsa = salsa;
        this.queso = queso;
        this.ingredientes = ingredientes;
        this.usuario = usuario;
        this.favorita = false;
        this.nombre = nombre;
        this.ref_img = R.drawable.pizza;
    }
    */

    public Pizza(Pizza pizza) {
        this.idPizza = pizza.getIdPizza();
        this.tamano = pizza.getTamano();
        this.salsa = pizza.getSalsa();
        this.queso = pizza.getQueso();
        this.ingredientes = pizza.getIngredientes();
        this.usuario = pizza.getUsuario();
        this.favorita = pizza.getFavorita();
        this.nombre = pizza.getNombre();
        DAOPizzas.getInstance().setPizzaFav(pizza, false);
        this.ref_img = pizza.getRef_img();
    }

    public Pizza(Integer idPizza) {
        this.idPizza = idPizza;
    }

    private Integer incrmentId(){
        return Pizza.contId++;
    }

    public Integer getIdPizza() {
        return idPizza;
    }

    public void setIdPizza(Integer idPizza) {
        this.idPizza = idPizza;
    }

    public Tamano getTamano() {
        return tamano;
    }

    public void setTamano(Tamano tamano) {
        this.tamano = tamano;
    }

    public Salsa getSalsa() {
        return salsa;
    }

    public void setSalsa(Salsa salsa) {
        this.salsa = salsa;
    }

    public Queso getQueso() {
        return queso;
    }

    public void setQueso(Queso queso) {
        this.queso = queso;
    }

    public ArrayList<Ingrediente> getIngredientes() {
        return ingredientes;
    }

    public void setIngredientes(ArrayList<Ingrediente> ingredientes) {
        this.ingredientes = ingredientes;
    }

    public Usuario getUsuario() {
        return usuario;
    }

    public void setUsuario(Usuario usuario) {
        this.usuario = usuario;
    }

    public Boolean getFavorita() {
        return favorita;
    }

    public void setFavorita(Boolean favorita) {
        this.favorita = favorita;
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public int getRef_img() {
        return ref_img;
    }

    public void setRef_img(int ref_img) {
        this.ref_img = ref_img;
    }

    public Double getPrecio() {
        Double precio = 0.0;

        precio += getTamano().getPrecioInicial();
        precio += getSalsa().getPrecioInicial();
        precio += getQueso().getPrecioInicial();

        for(int i = 0; i < getIngredientes().size(); i++){
            precio += getIngredientes().get(i).getPrecioInicial();
        }

        return precio;

    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Pizza pizza = (Pizza) o;
        return idPizza.equals(pizza.idPizza);
    }

    @Override
    public int hashCode() {
        return Objects.hash(idPizza);
    }

    @Override
    public String toString() {
        StringBuilder cadena = new StringBuilder();

        cadena.append("TAMA??O:\n\t - " + getTamano().getNombre() + "\n");
        cadena.append("SALSA:\n\t - " + getSalsa().getNombre() + "\n");
        cadena.append("QUESO:\n\t - " + getQueso().getNombre() + "\n");
        cadena.append("INGREDIENTES:\n");

        for(int i = 0; i < getIngredientes().size(); i++){
            cadena.append("\t - " + getIngredientes().get(i).getNombre() + "\n");
        }

        return cadena.toString();
    }

    public ContentValues toContentValues() {

        ContentValues values = new ContentValues();

        values.put(PizzaEntry.TAMANO, getTamano().ordinal());
        values.put(PizzaEntry.SALSA, getSalsa().ordinal());
        values.put(PizzaEntry.QUESO, getQueso().ordinal());
        values.put(UsuarioEntry.USUARIO, getUsuario().getUsuario());
        values.put(PizzaEntry.FAVORITA, getFavorita());
        values.put(PizzaEntry.NOMBRE, getNombre());

        return values;
    }

}

File: app/src/main/java/com/example/acdat_pizzeria/modelo/kotlin/Pizza.kt
package com.example.acdat_pizzeria.modelo.kotlin

import com.example.acdat_pizzeria.modelo.Usuario
import com.example.acdat_pizzeria.modelo.enums.Ingrediente
import com.example.acdat_pizzeria.modelo.enums.Queso
import com.example.acdat_pizzeria.modelo.enums.Salsa
import com.example.acdat_pizzeria.modelo.enums.Tamano
import java.io.Serializable

class Pizza : Serializable {

    companion object {
        private var contId = 0
    }

    private var idPizza: Int
    private lateinit var tamano: Tamano
    private lateinit var salsa: Salsa
    private lateinit var queso: Queso
    private lateinit var ingredientes: ArrayList<Ingrediente>
    private lateinit var usuario: Usuario
    private var favorita: Boolean? = null
    private lateinit var nombre: String

    constructor(
        tamano: Tamano,
        salsa: Salsa,
        queso: Queso,
        ingredientes: ArrayList<Ingrediente>,
        usuario: Usuario
    ) {
        idPizza = incrmentId()
        this.tamano = tamano
        this.salsa = salsa
        this.queso = queso
        this.ingredientes = ingredientes
        this.usuario = usuario
        favorita = false
        nombre = "Predeterminada"
    }

    constructor(
        salsa: Salsa,
        queso: Queso,
        ingredientes: ArrayList<Ingrediente>,
        nombre: String
    ) {
        idPizza = incrmentId()
        this.salsa = salsa
        this.queso = queso
        this.ingredientes = ingredientes
        favorita = false
        this.nombre = nombre
    }

    constructor(
        tamano: Tamano,
        salsa: Salsa,
        queso: Queso,
        ingredientes: ArrayList<Ingrediente>,
        usuario: Usuario,
        nombre: String
    ) {
        idPizza = incrmentId()
        this.tamano = tamano
        this.salsa = salsa
        this.queso = queso
        this.ingredientes = ingredientes
        this.usuario = usuario
        favorita = false
        this.nombre = nombre
    }

    constructor(pizza: Pizza) {
        idPizza = incrmentId()
        tamano = pizza.tamano
        salsa = pizza.salsa
        queso = pizza.queso
        ingredientes = pizza.ingredientes
        usuario = pizza.usuario
        favorita = pizza.favorita
        nombre = pizza.nombre
        pizza.favorita = false
    }

    constructor(idPizza: Int) {
        this.idPizza = idPizza
    }

    private fun incrmentId(): Int {
        return contId++
    }

    fun getIdPizza(): Int {
        return idPizza
    }

    fun setIdPizza(idPizza: Int) {
        this.idPizza = idPizza
    }

    fun getTamano(): Tamano {
        return tamano
    }

    fun setTamano(tamano: Tamano) {
        this.tamano = tamano
    }

    fun getSalsa(): Salsa {
        return salsa
    }

    fun setSalsa(salsa: Salsa) {
        this.salsa = salsa
    }

    fun getQueso(): Queso {
        return queso
    }

    fun setQueso(queso: Queso) {
        this.queso = queso
    }

    fun getIngredientes(): ArrayList<Ingrediente> {
        return ingredientes
    }

    fun setIngredientes(ingredientes: ArrayList<Ingrediente>) {
        this.ingredientes = ingredientes
    }

    fun getUsuario(): Usuario {
        return usuario
    }

    fun setUsuario(usuario: Usuario) {
        this.usuario = usuario
    }

    fun getFavorita(): Boolean? {
        return favorita
    }

    fun setFavorita(favorita: Boolean) {
        this.favorita = favorita
    }

    fun getNombre(): String {
        return nombre
    }

    fun setNombre(nombre: String) {
        this.nombre = nombre
    }

    val precio: Double
        get() {
            var precio = 0.0
            precio += tamano.precioInicial
            precio += salsa.precioInicial
            precio += queso.precioInicial
            for (i in ingredientes.indices) {
                precio += ingredientes[i].precioInicial
            }
            return precio
        }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Pizza

        if (idPizza != other.idPizza) return false

        return true
    }

    override fun hashCode(): Int {
        return idPizza
    }

    override fun toString(): String {
        val cadena = StringBuilder()
        cadena.append("TAMAÑO:\n\t - ${tamano.nombre}")
        cadena.append("SALSA:\n\t - ${salsa.nombre}")
        cadena.append("QUESO:\n\t - ${queso.nombre}")
        cadena.append("INGREDIENTES:\n")
        for (i in ingredientes!!.indices) {
            cadena.append("\t - ${ingredientes!![i].nombre}\n"
            )
        }
        return cadena.toString()
    }
}